zipline/_protocol.pyx

```diff
@@ -391,8 +399,74 @@ cdef class BarData:
             normalize_date(self.simulation_dt_func())
         )
 
+    #################
+    # OLD API SUPPORT
+    #################
+    cdef _calculate_universe(self):
+        if self._universe_func is None:
+            return []
+
+        simulation_dt = self.simulation_dt_func()
+        if self._last_calculated_universe is None or \
+                self._universe_last_updated_at != simulation_dt:
+
+            self._last_calculated_universe = self._universe_func()
+            self._universe_last_updated_at = simulation_dt
+
+        return self._last_calculated_universe
+
+    def __iter__(self):
+        self._warn_user("Iterating over the assets in `data` is "
+                        "deprecated.")
+        for asset in self._calculate_universe():
+            yield asset
+
+    def __contains__(self, asset):
+        self._warn_user("Checking whether an asset is in data is "
+                        "deprecated.")
+        universe = self._calculate_universe()
+        return asset in universe
+
+    def iteritems(self):
+        self._warn_user("Iterating over the assets in `data` is "
+                        "deprecated.")
+        for asset in self._calculate_universe():
+            yield asset, self[asset]
+
+    def __len__(self):
+        self._warn_user("Iterating over the assets in `data` is "
+                        "deprecated.")
+
+        return len(self._calculate_universe())
+
+    def keys(self):
+        self._warn_user("Iterating over the assets in `data` is "
+                        "deprecated.")
+
+        return list(self._calculate_universe())
+
+    def iterkeys(self):
+        iter(self.keys())
+
+    def __getitem__(self, name):
+        return self._get_equity_price_view(name)
+
+    cdef _warn_user(self, msg):
+        warnings.warn(
+            msg,
+            category=ZiplineDeprecationWarning,
+            stacklevel=1
```

Comment:
I have a test for this, it is what I want, because this is Cython code.  the first python stackframe is the algorithm code.